staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/gregjones/httpcache"
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+const (
+	oneDay                 = 24 * time.Hour
+	lastCompactionFilename = "lastcompactiontime"
+)
+
+// CompactableDiskCache implements the httpcache.Cache interface. It delegates to
+// sumDiskCache for Cache functions (Get, Set, Delete), and occasionally removes
+// old cached files. The specified "cacheSize" is NOT a hard limit; the disk
+// cache can breach this threshold. When cache compaction occurs, it will resize
+// the disk cache below this "cacheSize".
+//
+// Trigger:
+//
+//	disk cache compactions are triggered when a file is accessed and the
+//	last compaction occurred over a day ago. Every cache file access checks
+//	when the last compaction occurred by checking the modification time
+//	on an empty file in the cache -- "lastcompactiontime". File access
+//	time for a cached file is not used directly because it is not well
+//	supported by all OSes.
+//
+// Concurrency:
+//
+//	Within the same process: the goroutines of a single process are
+//	coordinated using "sync/singleflight" functionality.
+//
+//	Among different processes: if (for example) a separate kubectl command
+//	compacts the disk cache at the same time, separate processes can compete
+//	to delete the same files. This code must be tolerant to attempting to
+//	delete files already deleted.
+type CompactableDiskCache struct {
+	cacheDir  string
+	cacheSize int64
+	delegate  httpcache.Cache
+}
+
+// Get returns the []byte representation of a cached response and a bool
+// set to true if the value isn't empty. This function can trigger a
+// disk cache compaction which runs in a separate goroutine. Touches
+// cached file to update file modification to match access time.
+func (cdc *CompactableDiskCache) Get(key string) ([]byte, bool) {
+	// If the file exists, update file modification time = access time.
+	_ = touchFile(sanitize(key))
```

Comment:
Hm, I think the whole reason access time isn't implemented everywhere was to spare SSDs. This might add a lot of writes. Did we think about if that is OK?

(apologies if I asked for this, I thought I asked for touching the marker file but my memory could be wrong)

Reply:
Without the ability to make the access time equal the modification time, it really brings into question whether we can correctly evict the least recently used files from the cache. We can only consistently determine LRU through modification time. Maybe the best we can do is approximate LRU... or choose another strategy like randomly evicting files.

Reply:
> Hm, I think the whole reason access time isn't implemented everywhere was to spare SSDs. 

For performance reasons as well, to avoid file reads turning into writes. Probably not as big of a deal here.

Modern linux has multiple variants of atime tracking that are lighter weight and don't always immediately flush to disk (`relatime` and `lazyatime`) but I think mainly for performance reasons as SSDs have gotten more durable.

Related reading:
https://lwn.net/Articles/621046/
https://unix.stackexchange.com/a/646202

Have previously (ab)used atime [for Kubernetes's bazel build LRU cache](https://github.com/kubernetes/test-infra/blob/97064ce01a8838696e488138e690f203039fa182/greenhouse/gce-storage.yaml#L13) 🙃